src/auth/auth.cpp

```diff
@@ -232,72 +275,156 @@ auto ParseJson(std::string_view str) {
-
-    // Check if role is present
-    auto role = GetRole(rolename);
-    if (!role) {
-      spdlog::warn(utils::MessageWithLink("Couldn't authenticate user '{}' because the role '{}' doesn't exist.",
-                                          username, rolename, "https://memgr.ph/auth"));
+    const auto &ret_errors = ret.at("errors");
+    if (!ret_errors.is_string()) {
+      spdlog::warn(utils::MessageWithLink(
+          "Couldn't authenticate user: the external auth module did not pass the errors to Memgraph.",
+          "https://memgr.ph/sso"));
       return std::nullopt;
     }
+    const auto &errors = ret_errors.get<std::string>();
+    spdlog::warn(utils::MessageWithLink("Couldn't authenticate user:", errors, "https://memgr.ph/sso"));
+    return std::nullopt;
+  }
 
-    // Authenticate the user.
-    if (!is_authenticated) return std::nullopt;
-
-    return RoleWUsername{username, std::move(*role)};
+  if (is_authenticated && (ret.find("role") == ret.end() || ret.find("username") == ret.end())) {
+    spdlog::warn(utils::MessageWithLink(
+        "Couldn't authenticate user: the username/role fields were not returned by the external auth module.",
+        "https://memgr.ph/sso"));
+    return std::nullopt;
   }
 
-  /*
-   * LOCAL AUTH STORAGE
-   */
-  auto user = GetUser(username);
-  if (!user) {
-    spdlog::warn(utils::MessageWithLink("Couldn't authenticate user '{}' because the user doesn't exist.", username,
-                                        "https://memgr.ph/auth"));
+  const auto &ret_role = ret.at("role");
+  const auto &ret_username = ret.at("username");
+  if (!ret_username.is_string() || !ret_role.is_string()) {
+    spdlog::warn(
+        utils::MessageWithLink("Couldn't authenticate user: the username/role returned by the external auth module "
+                               "need to be string values.",
+                               "https://memgr.ph/sso"));
     return std::nullopt;
   }
-  if (!user->CheckPassword(password)) {
-    spdlog::warn(utils::MessageWithLink("Couldn't authenticate user '{}' because the password is not correct.",
-                                        username, "https://memgr.ph/auth"));
+
+  const auto &rolename = ret_role.get<std::string>();
+  const auto &username = ret_username.get<std::string>();
+
+  auto role = GetRole(rolename);
+  if (!role) {
+    spdlog::warn(utils::MessageWithLink("Couldn't authenticate external user because the role {} doesn't exist.",
+                                        rolename, "https://memgr.ph/auth"));
     return std::nullopt;
   }
-  if (user->UpgradeHash(password)) {
-    SaveUser(*user);
+  auto already_existing_user = GetUser(username);
+  if (already_existing_user) {
+    spdlog::warn(utils::MessageWithLink(
+        "Couldn't authenticate external user because a local user {} with the same name already exists.", rolename,
+        "https://memgr.ph/auth"));
+    return std::nullopt;
```

Comment:
Same comment as above

Reply:
From today’s Ante <-> Andrej call: login should fail if a SSO username has a duplicate local (native auth) username; this is in order to avoid audit log confusion.

Resolving this post-call